modules/puppet_proxy/puppet_cache.rb

```diff
@@ -1,100 +1,44 @@
-require 'json'
+require 'puppet_proxy/memory_store'
+require 'thread'
 
 module Proxy::Puppet
   class PuppetCache
-    extend Proxy::Log
-    class << self
-      def scan_directory_with_cache directory, environment, scanner
-        logger.info("Running scan_directory on #{environment}: #{directory}")
+    include Proxy::Log
 
-        cache = read_from_cache(environment)
+    attr_reader :class_scanner
 
-        if cache.has_key?(directory)
-          lcache = cache[directory]
-        else
-          lcache = {}
-        end
-
-        seenmodules=[]
-        changed    = false
-        manifest   = Dir.glob("#{directory}/*").map do |path|
-          puppetmodule = File.basename(path)
-          mtime        = File.mtime(path).to_i
-          seenmodules.push(puppetmodule)
-
-          lcache[puppetmodule] = {} unless lcache.has_key?(puppetmodule)
-          if lcache[puppetmodule].has_key?('timestamp') && lcache[puppetmodule]['timestamp'] >= mtime
-            logger.debug("Using cached class #{puppetmodule}")
-            modulemanifest = lcache[puppetmodule]['manifest']
-          else
-            changed = true
-            logger.info("Scanning class #{puppetmodule}")
-            modulemanifest = Dir.glob("#{path}/manifests/**/*.pp").map do |filename|
-              scanner.scan_manifest File.read(filename), filename
-            end
-
-            lcache[puppetmodule]['timestamp']= Time.new.to_i
-            lcache[puppetmodule]['manifest'] = modulemanifest
-          end
-          modulemanifest
-        end.compact.flatten
-
-        if changed
-          logger.info("Cache file need to be updated for #{environment}: #{directory}")
-          # Clean obsolete cache modules
-          oldlength = lcache.length
-          lcache.delete_if { |key, value| !seenmodules.include?(key) }
-          logger.info("Cleaning #{oldlength - lcache.length } modules from cache") if oldlength - lcache.length > 0
-
-          cache[directory] = lcache
-          write_to_cache(cache, environment)
-          logger.info("Cache file updated for #{environment}: #{directory}")
-        end
-
-        manifest
-      end
-
-      def read_from_cache environment
-        cachefile = File.expand_path("cache_#{environment}.json", Proxy::Puppet::Plugin.settings.cache_location)
+    def initialize(class_scanner, classes_store, timestamps_store)
+      @classes_cache = classes_store
+      @timestamps = timestamps_store
+      @class_scanner = class_scanner
+      @mutex = Mutex.new
```

Comment:
Isn't this going to create a new mutex per instance of the scanner, rather than one shared one across all threads?  (I assume the threads you're concerned about are the two possible rack/webrick instances.)